### ISSUE TITLE
Signed URL with Content-Disposition header should check RFC5987 encoding

### ISSUE THREAD

**Issue by viki53** [minio/minio#16159]
## Expected Behavior

When doing an API call to get a signed URL (using the S3 client from `aws-sdk` in Node.js) with some extra parameters I noticed a difference in behavior with the "real" S3 API.

Doing the same API call on AWS gives an error when Minio seems fine.

Looking at it closely, the issue is due to a `ResponseContentDisposition` parameter in the `getSignedUrlPromise('getObject', params)` call: AWS checks that the filename uses the encoding defined in [RFC 5987](https://datatracker.ietf.org/doc/html/rfc5987), but Minio doesn't.

## Current Behavior

When the filename contains some characters outside of that encoding (such as `Capture d’écran 2022-07-21 à 18.28.12.png`), AWS return an error on the URL returned by the endpoint, when Minio doesn't.

## Possible Solution

Check that the content of the `ResponseContentDisposition` (or ideally just the `filename` part) parameter matches the RFC5987 encoding.

## Steps to Reproduce (for bugs)

1. Make an API call on both AWS and Minio, like so (here using Node.js):
    ```ts
      s3Client
        .getSignedUrlPromise('getObject', {
          Key: 'some-key', // replace this with any valid value
          ResponseContentType: 'image/png',
          Bucket: 'your-bucket', // replace this with any valid value
          ResponseContentDisposition: `attachment; filename="Capture d’écran 2022-07-21 à 18.28.12.png"`,
        })
    ```
2. Visit the URL returned by this call
3. You should get an error on AWS
4. Minio currently doesn't return any error, even though the `filename` isn't valid

## Context

When testing locally with Minio this worked just fine, but once in staging/production using AWS it failed as behavior isn't the same.

## My Environment

```
minio version RELEASE.2022-10-29T06-21-33Z (commit-id=d765b89a63c225ca271ef6d031a7bcc8fe9137df)
Runtime: go1.19.2 darwin/arm64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2022 MinIO, Inc.
```

Running macOS 13.0.1

**Comment by harshavardhana** [comment]
> Looking at it closely, the issue is due to a `ResponseContentDisposition` parameter in the `getSignedUrlPromise('getObject', params)` call: AWS checks that the filename uses the encoding defined in [RFC 5987](https://datatracker.ietf.org/doc/html/rfc5987), but Minio doesn't.

I am not able to see this is any of the AWS documentation can you point me to the right direction?

**Comment by viki53** [comment]
I haven't looked at the AWS docs on this, but here's a sample error I got when visiting a signed URL, before encoding the `filename`:

```xml
<Error>
  <Code>InvalidArgument</Code>
  <Message>
    Header value cannot be represented using ISO-8859-1.
  </Message>
  <ArgumentName>response-content-disposition</ArgumentName>
  <ArgumentValue>
    attachment; filename="Capture d’écran 2022-07-21 à 18.28.12.png"
  </ArgumentValue>
  <RequestId>61RTX72MXNTXC2VR</RequestId>
  <HostId>
    5xmSPzzTrf28wnJb0b7etH9NC0o3t2glHmVFqWgP9Er+PfMct33M4NA73kNJvCOfDRBAYAGRE0M=
  </HostId>
</Error>
```

I found the explanation by looking at [the MDN docs for the `Content-Disposition` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) to be honest.

I didn't really expect AWS to do some extra check on this, but apparently they do.

**Comment by harshavardhana** [comment]
> I haven't looked at the AWS docs on this, but here's a sample error I got when visiting a signed URL, before encoding the `filename`:
> 
> ```
> <Error>
>   <Code>InvalidArgument</Code>
>   <Message>
>     Header value cannot be represented using ISO-8859-1.
>   </Message>
>   <ArgumentName>response-content-disposition</ArgumentName>
>   <ArgumentValue>
>     attachment; filename="Capture d’écran 2022-07-21 à 18.28.12.png"
>   </ArgumentValue>
>   <RequestId>61RTX72MXNTXC2VR</RequestId>
>   <HostId>
>     5xmSPzzTrf28wnJb0b7etH9NC0o3t2glHmVFqWgP9Er+PfMct33M4NA73kNJvCOfDRBAYAGRE0M=
>   </HostId>
> </Error>
> ```
> 
> I found the explanation by looking at [the MDN docs for the `Content-Disposition` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) to be honest.
> 
> I didn't really expect AWS to do some extra check on this, but apparently they do.

@viki53 what we support is full UTF-8, we could change this but this may cause some breakage for some applications.

**Comment by harshavardhana** [comment]
> I found the explanation by looking at the MDN docs for the Content-Disposition header to be honest.

The explanation here only talks about `filename*` to be encoded in the rfc you mentioned but not the `filename` parameter - this looks like some hidden server side restriction. Undocumented in AWS S3

**Comment by viki53** [comment]
I did a bit more digging: turns out `Content-Disposition` isn't really an official header, so browsers (or anyone, really) are free to do what they want. 

[RFC 2183](https://www.rfc-editor.org/rfc/rfc2183#section-2.3) acknowledged that `filename` was to be encoded in `US-ASCII` although that was judged too restrictive… back in 1997 😅 

Looking at [this StackOverflow answer](https://stackoverflow.com/a/20933751) the MDN doc might be a bit off: `filename*` can accept UTF-8 (or any specified encoding) but regular `filename` only accepts ISO-8859-1.

I also found [this article](https://engineering.resolvergroup.com/2022/02/aws-s3-utf-8-content-disposition/) that points to using `filename*` for UTF-8 characters, while still having a `filename` as fallback (with ISO-8859-1 characters only).
